rfcs/0167-maintainer-account.md

```diff
@@ -0,0 +1,270 @@
+6. *Wait at least one week for maintainer to respond*
+7. If maintainer does not still respond, the PR will be merged and the maintainer will be removed
+```
+
+## Update maintainers/maintainer-list.nix
+
+Update the preamble in [maintainers/maintainer-list.nix](https://github.com/NixOS/nixpkgs/blob/master/maintainers/maintainer-list.nix)
+
+````
+/* List of NixOS maintainers.
+    ```nix
+    handle = {
+      # Required
+      name = "Your name";
+      github = "GithubUsername";
+      githubId = your-github-id;
+
+      # **At least one** of email, matrix or discourse must be given in order to provide fallback communications
+      email = "address@example.org";
+      matrix = "@user:example.org";
+      discourse = "DiscourseUsername";
+
+      # Optional
+      keys = [{
+        fingerprint = "AAAA BBBB CCCC DDDD EEEE  FFFF 0000 1111 2222 3333";
+      }];
+    };
+    ```
+
+    where
+
+    - `handle` is the handle you are going to use in nixpkgs expressions, strongly preferred to match `github`
+    - `name` is a name that people would know and recognize you by,
+    - `github` is your GitHub handle (as it appears in the URL of your profile page, `https://github.com/<userhandle>`),
+    - `githubId` is your GitHub user ID, which can be found at `https://api.github.com/users/<userhandle>`,
+    - `email` is your maintainer email address,
+    - `matrix` is your Matrix user ID,
+    - `discourse` is your `https://discourse.nixos.org` username,
+    - `keys` is a list of your PGP/GPG key fingerprints.
+
+    # Editing
+
+    When editing this file:
+     * keep the list alphabetically sorted, check with:
+         nix-instantiate --eval maintainers/scripts/check-maintainers-sorted.nix
+     * test the validity of the format with:
+         nix-build lib/tests/maintainers.nix
+
+    See `./scripts/check-maintainer-github-handles.sh` for an example on how to work with this data.
+
+    # GitHub username
+
+    Maintainers must maintain an active account on the primary nixpkgs hosting platform.
+    GitHub is currently the primary nixpkgs hosting platform, so you are required to provide and keep active a GitHub account.
+
+    This information ensures that you:
+    - Can actively participate in Issues and Pull Requests
+    - Get invited to the @NixOS/nixpkgs-maintainers team
+    - Are reachable by mention on Issues and Pull Requests, either by a human or a robot.
+
+    `handle == github` is preferred whenever the username is an acceptable attribute name.
+
+    If `github` begins with a numeral, `handle` should be prefixed with an underscore.
+    ```nix
+    _1example = {
+      github = "1example";
+    };
+    ```
+
+    # Alternative form of contact
+
+    Maintainers *must* provide at least one of the alternative forms of contact. The simplest option would be to provide
```

Comment:
```suggestion
    Maintainers SHOULD provide at least one of the alternative forms of contact. The simplest option would be to provide
```
I disagree that maintainers MUST provide this info. As long as the maintainers are accessible on github, this is a non-issue; these are fallback forms of contact anyway, and they can be onerous requirements that serve to increase the barrier to contribution. I'm fine with "SHOULD".

(definitions are as per [IETF RFC 2119](https://www.rfc-editor.org/rfc/rfc2119).)

Reply:
I cannot understand your objection, there is literally a trivial email address that is linked to your GitHub account that is public and already usable.

If you are saying that people put null email addresses in their GitHub account, I am not so sure if we have to support the onerous requirement for them to have a working way of being contacted over email.

Reply:
The `MUST` is trivially fulfilled here.

Reply:
> I cannot understand your objection

Anonymity or, quite simply, not wanting to be contacted elsewhere. If maintainers don't respond to the GH notification, remove them as per the process already outlined.

> there is literally a trivial email address that is linked to your GitHub account that is public and already usable.

Untrue, gh provides anonymised, no-reply "committer" emails.

Reply:
@eclairevoyant Indeed, I missed that, but then, this is a social project, if you cannot afford to be present on any other platform beyond GitHub and want to be extremely anonymous, I think it will be very hard to be a maintainer, but that's my personal opinion.

Is anonymity a possibility we want to offer for the model of maintainers we want?

Reply:
> if you cannot afford to be present on any other platform beyond GitHub ... I think it will be very hard to be a maintainer

That contradicts a point elsewhere in the RFC regarding the downsides of using email:

> Users and maintainers must leave the nixpkgs hosting platform in order to contact a maintainer, possibly fracturing communications and losing public history

Package or module discussions should stay public; everyone should be able to benefit from the conversation. This is not an encouragement to be antisocial; on the contrary, it's further cementing the expectation that maintainers use and respond via gh if they want to stay as maintainers.

If the maintainer is unresponsive, it also reduces the burden of the "initiator" to remove the maintainer, as the initiator only needs to reach out on one platform (rather than follow up on an email as well).

Reply:
> That contradicts a point elsewhere in the RFC regarding the downsides of using email:

> > Users and maintainers must leave the nixpkgs hosting platform in order to contact a maintainer, possibly fracturing communications and losing public history

It doesn't contradict each other. Yes reviews should be done over github so things are transparent.
But nix is a community project so having other ways to communicate feels like a good requirement, for example before I nominated people for commit bit I talked with them in private which wouldn't really be possible via github. And a lot of other things like discussing rfc implementation progress, developing new tools, talking about ideas happens via matrix or sometimes email, so having that as a requirement sounds reasonable.

Reply:
@eclairevoyant The problem is that you are having what I feel like is a virtual or an abstract concern about people not being able to provide an email address, and we are debating over that virtual concern.

Conversely, as a committer, as a release manager, I had multiple situations where I was not able to reach out to folks because they didn't provide a secondary contact point and that pinging them on GitHub was not efficient.

Maintainership has not been in very good shape, so I'd ask you to provide a little data panel of how many maintainers we have without a secondary contact point and what are the packages they are maintaining, other than that, I don't feel compelled to act upon those comments. Feel free to join the shepherd team if you care about this.

Reply:
> you are having what I feel like is a virtual or an abstract concern

There it is. Well, what I see is that you are replying to every concern with dismissiveness, and are therefore _assuming_ this is an abstract concern.

> I'd ask you to provide a little data panel of how many maintainers we have without a secondary contact point

If >99% of people have an email, then it proves that making it a requirement (MUST) won't help get maintainership in "good shape", and making it a recommendation (SHOULD) would have the same effect. If <1% of people have an email, then SHOULD can still improve the situation, and MUST is overkill. Either way, this data has nothing to do with my point.

Going back to the original motivation mentioned in the RFC:

> For example, a maintainer may choose not to provide their user information for the primary nixpkgs hosting platform.
> This opting out for one or more individuals can put extra maintenance overhead on the remaining maintainers; it fragments information which may be important for the overall community; and may even hide information from others working on nixpkgs.

It's quite clear that having a GH account, and setting up notifications accordingly, is necessary to be a maintainer on nixpkgs.

What I don't see is how having an email is "more efficient" when it takes longer to write an email than a GH comment, and requires a separate lookup to get the actual email (vs @-ing the gh username, which should almost always match the attrname in `lib.maintainers`)?

This is not a "virtual concern", it's a boundary. Nixpkgs is not every maintainer's life, nor does it necessarily have to be. If you choose to set your boundaries by saying "contact me on GH or these dozen other platforms" and be bombarded with this, then so be it, but forcing every maintainer to do so is both unreasonable and does not serve the purpose of ensuring maintainers are reachable. The main place we expect reachability for the responsibilities of maintainership is GH, so GH is the only MUST here. I do not want anonymous users bombarding my personal email _and_ GH just because some package build failed; let the notifications be in the **one public place** where maintainers have to be anyway.

Reply:
> There it is. Well, what I see is that you are replying to every concern with dismissiveness, and are therefore assuming this is an abstract concern.

Well, yes, I am a bit tired of having the same things happening in all of my RFCs and debating what I feel like is non-concern at the scale of things we are operating.

> If >99% of people have an email, then it proves that making it a requirement (MUST) won't help get maintainership in "good shape", and making it a recommendation (SHOULD) would have the same effect. If <1% of people have an email, then SHOULD can still improve the situation, and MUST is overkill. Either way, this data has nothing to do with my point.

It proves nothing? There was no RFC before. If anything, this was all triggered by people proving they want clear policy on the matter because not having policy was deemed insufficient.

> It's quite clear that having a GH account, and setting up notifications accordingly, is necessary to be a maintainer on nixpkgs.

Interesting, it doesn't seem like everyone share this perspective, as this RFC was spawned off the drama of someone threatening GDPR enforcement because a GH account should not be required to contribute.

May I recommend reading up on the context surrounding this PR rather than doing assumptions?

> What I don't see is how having an email is "more efficient" when it takes longer to write an email than a GH comment, and requires a separate lookup to get the actual email (vs @-ing the gh username, which should almost always match the attrname in lib.maintainers)?

Because sometimes GitHub does not work or because we need to send private information (e.g. CVE handling when we do have embargoes, etc.).
It also acts as a backup if we don't have GitHub for any reason.

The "taking longer to write an email" than "a GH comment" is astounding to me, how do you know? I write emails faster than GH comments, given the brokenness of the GitHub UI.

(which should almost always match the attrname means that I have to lookup lib.maintainers and not once or twice, as a release manager BTW, so that's moot).

> This is not a "virtual concern", it's a boundary. Nixpkgs is not every maintainer's life, nor does it necessarily have to be. If you choose to set your boundaries by saying "contact me on GH or these dozen other platforms" and be bombarded with this, then so be it, but forcing every maintainer to do so is both unreasonable and does not serve the purpose of ensuring maintainers are reachable. 

Yes but this boundary you are imposing is something you want to see, for it to happen, you need to participate in the RFC process and work with the shepherds, if all you propose is to order us around to execute on your concerns, I am sorry this is not how RFCs work.

Yes, Nixpkgs is not every maintainer's life and I do not intend to make it so by asking two point of contact to every person who decide to take care of a package, otherwise, please make it clear how asking for email makes Nixpkgs the maintainer's life. Many other distroes have the **email** of their maintainer, I fail to understand how we are very different to them?

If your concern is about being bombarded by email as soon as you put the email, please consider creating an email address for contribution purpose, a lot of us do that, and it's not very hard, sometimes much smaller than maintaining packages. To be clear, spamming is a valid concern and if it came to our knowledge that someone is abusing those email addresses, that'd be a big problem.
In the meantime, I am maintainer of a bunch of things and my email is on there, I received 2 emails in one year regarding package maintenance, other than that, nothing. Maybe my experience is not representative though, maybe I don't maintain enough packages or high-profile packages, but here we go.

> The main place we expect reachability for the responsibilities of maintainership is GH, so GH is the only MUST here. I do not want anonymous users bombarding my personal email and GH just because some package build failed; let the notifications be in the one public place where maintainers have to be anyway.

Sure, I don't want to argue further with you in comments. I propose you two choices:

- Join the shepherd team and work with us regarding your concerns and my concerns and let's find an acceptable solution
- Refuse, but in that case, I am afraid I cannot action your comments

Reply:
> May I recommend reading up on the context surrounding this PR rather than doing assumptions?

I did read that thread, in its entirety, before responding here. I should have said "it's quite clear _to me_" as I assumed that was implied from context; nothing has universal agreement.

I'm not "ordering you around", that's a pretty bad-faith response to my comments. Anyway, if you want to be dismissive, so be it, I'll save my time. I'm not going to spend my free time setting up some Matrix server just to repeat the same things I said here, when I already spend enough time on nixpkgs.

Reply:
> I'm not "ordering you around", that's a pretty bad-faith response to my comments. Anyway, if you want to be dismissive, so be it, I'll save my time. I'm not going to spend my free time setting up some Matrix server just to repeat the same things I said here, when I already spend enough time on nixpkgs.

Get me right, I don't want to be dismissive of your comments, but I spend also enough of my personal time in the Nix ecosystem (almost full-time?), the RFC process is that if this is a blocker for you, you can join the shepherd team, and we can have a call about this and try to work out the points.

If we cannot have this, I am sorry, but I had this in many of my RFCs and it's indistinguishable of me having to jump and have someone tell me "I want this, and I don't want to join the team to work with you", this is my definition of being ordered around. If this was not your intent, I conjure you to join the shepherd team, it doesn't require a Matrix server, AFAIK.

Reply:
>  It also acts as a backup if we don't have GitHub for any reason.

In my opinion, this is the primary reason to have a second form of contact. I do not find it wise to solely rely on GitHub. It exists today and works for us, but if it doesn't tomorrow then I want the Nix/NixOS community to have a way to gather back together elsewhere.

> What I don't see is how having an email is "more efficient" when it takes longer to write an email than a GH comment, and requires a separate lookup to get the actual email (vs @-ing the gh username, which should almost always match the attrname in lib.maintainers)?

This sentence you're discussing actually doesn't even require email. It requires _one_ of the alternatives. Though the more this has been discussed, the more convinced I am that it should just be email and discourse/matrix should be dropped as fulfilling the second contact method.

> Is anonymity a possibility we want to offer for the model of maintainers we want?

I don't think so.

Reply:
I've opted to not make email the only alternative. If others want to advocate, I'm happy to change my mind.